### PR TITLE
bindinfo: extract the table hint from the union statement

### LINKED ISSUE #50068: planner: set_var cannot take effect in UNION ALL statements
## Bug Report

Please answer these questions before submitting your issue. Thanks!

### 1. Minimal reproduce step (Required)

```
create table t (a int);
insert into t values (1);
insert into t values (1);

mysql> select /*+ set_var(max_execution_time=100) */ a, sleep(1) from t;
ERROR 3024 (HY000): Query execution was interrupted, maximum statement execution time exceeded

mysql> select /*+ set_var(max_execution_time=100) */ a, sleep(1) from t union all select a, 1 from t;
+------+----------+
| a    | sleep(1) |
+------+----------+
|    1 |        1 |
|    1 |        1 |
|    1 |        0 |
|    1 |        0 |
+------+----------+
4 rows in set (2.01 sec)
```

### 2. What did you expect to see? (Required)

The second query should return an error.

### 3. What did you see instead (Required)

The second query's execution time exceeds the limitation.

The reason is that UNION ALL's node is not considered when extracting statement hints:

img_v3_026p_32b954e0-92b3-4a15-8aac-c847fca3026g

### 4. What is your TiDB version? (Required)

Master

### PR DESCRIPTION
**Pull request by hawkingrei**
### What problem does this PR solve?

Issue Number: close #50068

Problem Summary:

### Check List

Tests 

- [x] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No need to test
  > - [ ] I checked and no code files have been changed.
  > 

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
None
```

### CODE REVIEW

**Review by qw4990 — APPROVED** [review]

File: pkg/util/hint/hint_processor.go [hunk 1]
```diff
@@ -88,6 +88,18 @@ func ExtractTableHintsFromStmtNode(node ast.Node, sctx sessionctx.Context) []*as
 		// check duplicated hints
 		checkInsertStmtHintDuplicated(node, sctx)
 		return x.TableHints
+	case *ast.SetOprStmt:
+		var result []*ast.TableOptimizerHint
```
Comment (hi-rustin) [inline pkg/util/hint/hint_processor.go:92]: Do we need to preallocate here? Might it not be worth it because usually we only have a few hints?

**Review by hi-rustin — APPROVED** [review]

File: pkg/util/hint/hint_processor.go [hunk 2]
```diff
@@ -88,6 +88,18 @@ func ExtractTableHintsFromStmtNode(node ast.Node, sctx sessionctx.Context) []*as
...
+		if x.SelectList == nil {
+			return nil
+		}
+		for _, s := range x.SelectList.Selects {
+			tmp := ExtractTableHintsFromStmtNode(s, sctx)
```
Comment (hi-rustin) [inline pkg/util/hint/hint_processor.go:97]: Maybe we can call it hints?

**Review by hi-rustin — APPROVED** [review]

File: pkg/util/hint/hint_processor.go (hunk 1 above)
Comment (hawkingrei) [reply pkg/util/hint/hint_processor.go:92]: It‘s hard to know how much memory do we need to allocate. 

BTW, Many statements are without hint so preallocate is unnecessary.

**Review by elsa0520 — APPROVED** [review]
LGTM

**Review by AilinKid — APPROVED** [review]

**Pull request merged**